### PR TITLE
Wiring up GPGX as core for SMS, GG, and SG

### LINKED ISSUE #3214: [Request] Add Genesis Plus GX's SMS core
Genesis Plus GX's SMS core probably has the best emulation available for the system. The emulator claims 100% compatibility on all platforms it supports, which is probably true as I've never ran into any SMS bugs while using it.

The reason I'm requesting this is that SMSHawk is problematic. Until recently several games were bugged, those got fixed but it broke some stuff elsewhere. An example is issue #3187, which gives the impression the core is some sort of house of cards similar to PCEHawk. Another problem is that audio emulation is kind of bad. Originally it was too quiet before the #2814 fix, but that ended up making the FM a little too grating. Now there are some nasty popping sounds in some games (examples are Phantasy Star's town or dungeon themes). Those sound much more clean on GPGX.

By the way, that SMS core already works through libretro. I even made a test movie with one game (Shinobi) and it synced perfectly, so save states seem stable. Since the hawk already has all components necessary to run it, maybe it'd not be that hard to integrate it properly? (SMSHawk doesn't need to be deleted, we could have two options similar to Nyma/PCEHawk.)

### LINKED ISSUE #3896: Wire up Game Gear, Master System, and SG-1000 support in Genplus-gx
The benefit of this would be having an additional, well tested core (GPGX) for these two systems.

The core is being implemented already #3889.  We'd only need to add BK core changes to direct GG, SG1000, SMS games to it.

### PR DESCRIPTION
**Pull request by SergioMartin86**
[//]: # "This description supports Markdown syntax. There's a cheatsheet here: https://guides.github.com/features/mastering-markdown/"
[//]: # "These lines are comments, for letting you know what you should be writing. You can delete them or leave them in."
[//]: # "Also, please remember to link related Issues! If a bug hasn't been reported, you may submit a fix without creating an Issue."

Enabling the recently updated GPGX core to be an option for emulating Sega Master System,  Game Gear, and SG-1000 games. This PR resolves #3214 and resolves #3896.

[//]: # "Apart from the mandatory license signature, these tasks are optional, but doing them could save reviewers some time and get the PR merged sooner."
Check if completed:
- [x] I have run any relevant test suites
- [x] I, the committer, have read the [licensing terms for contributors](https://github.com/TASEmulators/BizHawk/blob/master/contributing.md#copyrights-and-licensing) (last updated 2024-03-20) and am compliant

Tasks:

- [x] Add GPGX to the core list for SMS and GG
- [x]  Ensure it loads the corresponding BIOS correctly (GG)
- [x]  Ensure it loads the corresponding BIOS correctly (SMS)
- [x] Ensure it works correctly and syncs to native (GG)
- [x] Ensure it works correctly and syncs to native (SMS)
- [x] Ensure it works correctly and syncs to native (SG-1000)
- [x] Personalize configuration (especially sound) to SMS/GG/SG
- [x] Personalize controllers
- [x] Detect lag input

### CODE REVIEW

**Comment by YoshiRulz** [comment]
re: controllers, it's probably not worth trying to re-use SMSHawk code, except maybe the Virtual Pads.

File: src/BizHawk.Client.Common/config/Config.cs [hunk 1]
```diff
@@ -36,10 +36,18 @@ public class Config
...
+			(new[] { VSystemID.Raw.GG },
+				new[] { CoreNames.Gpgx, CoreNames.SMSHawk }),
 			(new[] { VSystemID.Raw.PCE, VSystemID.Raw.PCECD, VSystemID.Raw.SGX, VSystemID.Raw.SGXCD },
 				new[] { CoreNames.TurboNyma, CoreNames.HyperNyma, CoreNames.PceHawk }),
 			(new[] { VSystemID.Raw.PSX },
 				new[] { CoreNames.Octoshock, CoreNames.Nymashock }),
+			(new[] { VSystemID.Raw.SMS },
+				new[] { CoreNames.Gpgx, CoreNames.SMSHawk }),
+			(new[] { VSystemID.Raw.SG },
+				new[] { CoreNames.Gpgx, CoreNames.SMSHawk }),
```
Comment (YoshiRulz) [inline src/BizHawk.Client.Common/config/Config.cs:50]: You can combine all these, as GB and PCE are.

File: src/BizHawk.Emulation.Cores/Consoles/Sega/SMS/SMS.cs [hunk 2]
```diff
@@ -19,9 +19,9 @@ namespace BizHawk.Emulation.Cores.Sega.MasterSystem
 	public partial class SMS : IEmulator, ISoundProvider, ISaveRam, IInputPollable, IRegionable,
 		IDebuggable, ISettable<SMS.SmsSettings, SMS.SmsSyncSettings>, ICodeDataLogger
 	{
-		[CoreConstructor(VSystemID.Raw.SMS)]
-		[CoreConstructor(VSystemID.Raw.SG)]
-		[CoreConstructor(VSystemID.Raw.GG)]
+		[CoreConstructor(VSystemID.Raw.SMS, Priority = CorePriority.Low)]
+		[CoreConstructor(VSystemID.Raw.SG, Priority = CorePriority.Low)]
+		[CoreConstructor(VSystemID.Raw.GG, Priority = CorePriority.Low)]
```
Comment (YoshiRulz) [inline src/BizHawk.Emulation.Cores/Consoles/Sega/SMS/SMS.cs:24]: I don't think this is necessary? I've forgotten when exactly `Priority` has an effect, but it's only in some edge case or other.

Comment (CasualPokePlayer) [reply src/BizHawk.Emulation.Cores/Consoles/Sega/SMS/SMS.cs:24]: Priority is only intended to ever interact with fallback logic. If only two cores are present, it is meaningless.

File: src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs [hunk 3]
```diff
@@ -26,7 +30,11 @@ public GPGX(CoreLoadParameters<GPGXSettings, GPGXSyncSettings> lp)
 
 			ServiceProvider = new BasicServiceProvider(this);
 			// this can influence some things internally (autodetect romtype, etc)
+
 			string romextension = "GEN";
+			if (lp.Roms.FirstOrDefault()?.Game.System == VSystemID.Raw.SMS) romextension = "SMS";
+			if (lp.Roms.FirstOrDefault()?.Game.System == VSystemID.Raw.GG) romextension = "GG";
+			if (lp.Roms.FirstOrDefault()?.Game.System == VSystemID.Raw.SG) romextension = "SG";
```
Comment (YoshiRulz) [inline src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs:37]:
```suggestion
			var romextension = lp.Roms.FirstOrDefault()?.Game?.System switch
			{
				VSystemID.Raw.GG => "GG",
				VSystemID.Raw.SG => "SG",
				VSystemID.Raw.SMS => "SMS",
				_ => "GEN"
			};
```

File: src/BizHawk.Client.Common/config/Config.cs (hunk 1 above)
Comment (SergioMartin86) [reply src/BizHawk.Client.Common/config/Config.cs:50]: I tried that, but then you cannot individually select it per console.

**Comment by SergioMartin86** [comment]
No longer a draft, these three consoles are now running correctly under GPGX.

I will take care of the three remaining tasks (ensure sync)  in the meantime.

File: src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.ISettable.cs [hunk 4]
```diff
@@ -263,7 +263,17 @@ public class GPGXSyncSettings
...
 			public LibGPGX.Region Region { get; set; }
 
-			[DisplayName("FM Sound Chip Type")]
+			[DisplayName("[SMS/GG] Load BIOS")]
+			[Description("Indicates whether to load the system BIOS rom.")]
+			[DefaultValue(false)]
+			public bool loadBIOS { get; set; }
+
+			[DisplayName("[SMS/GG] FM Sound Chip Type")]
+			[Description("Sets the method used to emulate the FM synthesizer (main sound generator) of the SMS/GG.  'MAME' options are fast, and run full speed on most systems.  'Nuked' options are cycle accurate, very high quality, and have substantial CPU requirements.  The 'YM2413' chip is used by the original Sega Master System")]
```
Comment (vadosnaprimer) [inline src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.ISettable.cs:272]: > FM synthesizer (main sound generator) of the SMS/GG

GameGear didn't have it.

> The 'YM2413' chip is used by the original Sega Master System

Source?

**Comment by SergioMartin86** [comment]
Note: I'm done with all the functional changes. Should there remain any inaccuracies or things to adjust for the code to be more adequate to BizHawk please go ahead and fix them directly.

File: src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.ISettable.cs [hunk 5]
```diff
@@ -263,7 +263,17 @@ public class GPGXSyncSettings
...
 
-			[DisplayName("FM Sound Chip Type")]
+			[DisplayName("[SMS/GG] Load BIOS")]
+			[Description("Indicates whether to load the system BIOS rom.")]
+			[DefaultValue(false)]
+			public bool loadBIOS { get; set; }
+
+			[DisplayName("[SMS] FM Sound Chip Type")]
+			[Description("Sets the method used to emulate the FM synthesizer (main sound generator) of the SMS. 'MAME' is fast and runs full speed on most systems.'Nuked' is cycle accurate, very high quality, and have substantial CPU requirements.")]
+			[DefaultValue(LibGPGX.InitSettings.SMSFMSoundChipType.YM2413_MAME)]
```
Comment (vadosnaprimer) [inline src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.ISettable.cs:273]: FM synth is not the main sound generator on SMS. The correct first sentence would be `Sets the method used to emulate the FM Sound Unit of the Sega Mark III/Master System.`

File: waterbox/gpgx/cinterface/cinterface.c [hunk 6]
```diff
@@ -658,54 +660,53 @@ GPGX_EX int gpgx_init(const char* feromextension,
...
+		  config.ym2413 = 1;
+    break;
 
 	  case YM2413_NUKED: 
-	    config.opll = 1;
-		config.ym2413 = 0;
-        break;
+	   config.opll = 1;
+		  config.ym2413 = 0;
+    break;
```
[the remark this answers is not in the document]
Comment (vadosnaprimer) [reply waterbox/gpgx/cinterface/cinterface.c]: Setting ym2413 to 1 here is my final request I guess.

**Comment by SergioMartin86** [comment]
AFAIK, there is nothing else to do here. This is merge-ready

File: src/BizHawk.Client.Common/config/Config.cs (hunk 1 above)
Comment (YoshiRulz) [reply src/BizHawk.Client.Common/config/Config.cs:50]: Is that desired? I thought the whole point was to use GPGX' more accurate emulation by default; most people won't be switching at all, and if they did I don't see why they'd want it to be separate for the sister consoles.

File: src/BizHawk.Emulation.Cores/Consoles/Sega/SMS/SMS.cs (hunk 2 above)
Comment (YoshiRulz) [reply src/BizHawk.Emulation.Cores/Consoles/Sega/SMS/SMS.cs:24]: (read: this is not necessary and can be reverted)

File: src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs [hunk 7]
```diff
@@ -8,14 +8,19 @@
...
-	[PortedCore(CoreNames.Gpgx, "Eke-Eke", "25a90c6", "https://github.com/ekeeke/Genesis-Plus-GX")]
+	[PortedCore(CoreNames.Gpgx, "Eke-Eke", "0c45a8a", "https://github.com/ekeeke/Genesis-Plus-GX")]
 	public partial class GPGX : IEmulator, IVideoProvider, ISaveRam, IStatable, IRegionable,
 		IInputPollable, IDebuggable, IDriveLight, ICodeDataLogger, IDisassemblable
 	{
 		[CoreConstructor(VSystemID.Raw.GEN)]
+		[CoreConstructor(VSystemID.Raw.SMS)]
+		[CoreConstructor(VSystemID.Raw.GG)]
+		[CoreConstructor(VSystemID.Raw.SG)]
+
```
Comment (YoshiRulz) [inline src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs:23]: Random blank line

File: src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs [hunk 8]
```diff
@@ -8,14 +8,19 @@
 using BizHawk.Common;
 using BizHawk.Emulation.DiscSystem;
 using System.Linq;
+using System.IO;
 
 namespace BizHawk.Emulation.Cores.Consoles.Sega.gpgx
 {
-	[PortedCore(CoreNames.Gpgx, "Eke-Eke", "25a90c6", "https://github.com/ekeeke/Genesis-Plus-GX")]
+	[PortedCore(CoreNames.Gpgx, "Eke-Eke", "0c45a8a", "https://github.com/ekeeke/Genesis-Plus-GX")]
```
Comment (YoshiRulz) [inline src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs:15]:
```suggestion
	[PortedCore(
		name: CoreNames.Gpgx,
		author: "Eke-Eke",
		portedVersion: "0c45a8a",
		portedUrl: "https://github.com/ekeeke/Genesis-Plus-GX")]
```

File: src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGXControlConverter.cs [hunk 9]
```diff
@@ -215,10 +236,15 @@ public GPGXControlConverter(LibGPGX.InputData input, bool cdButtons)
 						player++;
 						break;
 					case LibGPGX.INPUT_DEVICE.DEVICE_PAD2B:
+						if (systemId == VSystemID.Raw.SMS) AddToController(i, player, SMS2B);
+						if (systemId == VSystemID.Raw.SG) AddToController(i, player, GameGear);
+						if (systemId == VSystemID.Raw.GG) AddToController(i, player, GameGear);
```
Comment (YoshiRulz) [inline src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGXControlConverter.cs:241]:
```suggestion
						AddToController(i, player, systemId is VSystemID.Raw.SMS ? SMS2B : GameGear);
```

File: src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs [hunk 10]
```diff
@@ -365,7 +380,7 @@ private void SetControllerDefinition()
 			if (!Core.gpgx_get_control(input, inputsize))
 				throw new Exception($"{nameof(Core.gpgx_get_control)}() failed");
 
-			ControlConverter = new GPGXControlConverter(input, _cds != null);
+			ControlConverter = new GPGXControlConverter(input, SystemId, _cds != null);
```
Comment (YoshiRulz) [inline src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs:383]:
```suggestion
			ControlConverter = new GPGXControlConverter(input, systemId: SystemId, cdButtons: _cds is not null);
```

File: src/BizHawk.Client.Common/movie/import/GmvImport.cs [hunk 11]
```diff
@@ -88,7 +88,7 @@ protected override void RunImport()
 					: LibGPGX.INPUT_DEVICE.DEVICE_PAD3B;
 			}
 
-			var controlConverter = new GPGXControlConverter(input, false);
+			var controlConverter = new GPGXControlConverter(input, VSystemID.Raw.GEN, false);
```
Comment (YoshiRulz) [inline src/BizHawk.Client.Common/movie/import/GmvImport.cs:91]:
```suggestion
			GPGXControlConverter controlConverter = new(input, systemId: VSystemID.Raw.GEN, cdButtons: false);
```

File: src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs [hunk 12]
```diff
@@ -26,8 +31,13 @@ public GPGX(CoreLoadParameters<GPGXSettings, GPGXSyncSettings> lp)
 
 			ServiceProvider = new BasicServiceProvider(this);
 			// this can influence some things internally (autodetect romtype, etc)
-			string romextension = "GEN";
 
+			// Determining system ID from the rom. If no rom provided, assume Genesis (Sega CD)
+			SystemId = lp.Roms.Any() ? lp.Roms[0].Game.System : VSystemID.Raw.GEN;
+
+			// We need to pass the exact file extension to GPGX for it to correctly interpret the console
+			string RomExtension = lp.Roms.Any() ? Path.GetExtension(lp.Roms[0].RomPath).Replace(".", "") : "";
```
Comment (YoshiRulz) [inline src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs:39]:
```suggestion
			// Determining system ID from the rom. If no rom provided, assume Genesis (Sega CD)
			SystemId = VSystemID.Raw.GEN;
			var RomExtension = string.Empty;
			if (lp.Roms.Count >= 1)
			{
				SystemId = lp.Roms[0].Game.System;
				// We need to pass the exact file extension to GPGX for it to correctly interpret the console
				RomExtension = Path.GetExtension(lp.Roms[0].RomPath).RemovePrefix('.');
			}
```
also don't use PascalCase for locals

File: src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs [hunk 13]
```diff
@@ -228,19 +238,24 @@ private int load_archive(string filename, IntPtr buffer, int maxsize)
...
-					case "CD_BIOS_EU": firmwareID = "CD_BIOS_EU"; break;
-					case "CD_BIOS_JP": firmwareID = "CD_BIOS_JP"; break;
-					case "CD_BIOS_US": firmwareID = "CD_BIOS_US"; break;
+					case "CD_BIOS_EU": system = VSystemID.Raw.GEN; firmwareID = "CD_BIOS_EU"; break;
+					case "CD_BIOS_JP": system = VSystemID.Raw.GEN; firmwareID = "CD_BIOS_JP"; break;
+					case "CD_BIOS_US": system = VSystemID.Raw.GEN; firmwareID = "CD_BIOS_US"; break;
+					case "GG_BIOS":    system = VSystemID.Raw.SMS; firmwareID = "Japan";    break;
+					case "MS_BIOS_EU": system = VSystemID.Raw.SMS; firmwareID = "Export"; break;
+					case "MS_BIOS_JP": system = VSystemID.Raw.SMS; firmwareID = "Japan"; break;
+					case "MS_BIOS_US": system = VSystemID.Raw.SMS; firmwareID = "Export"; break;
```
Comment (YoshiRulz) [inline src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs:251]: Ugh GitHub doesn't like this one.
```diff
-				string system = null;
-				string firmwareID = null;
-				switch (filename)
-				{
-					case "CD_BIOS_EU": system = VSystemID.Raw.GEN; firmwareID = "CD_BIOS_EU"; break;
-					case "CD_BIOS_JP": system = VSystemID.Raw.GEN; firmwareID = "CD_BIOS_JP"; break;
-					case "CD_BIOS_US": system = VSystemID.Raw.GEN; firmwareID = "CD_BIOS_US"; break;
-					case "GG_BIOS":    system = VSystemID.Raw.SMS; firmwareID = "Japan";    break;
-					case "MS_BIOS_EU": system = VSystemID.Raw.SMS; firmwareID = "Export"; break;
-					case "MS_BIOS_JP": system = VSystemID.Raw.SMS; firmwareID = "Japan"; break;
-					case "MS_BIOS_US": system = VSystemID.Raw.SMS; firmwareID = "Export"; break;
-					default:
-						break;
-				}
+				FirmwareID? firmwareID = filename switch
+				{
+					"CD_BIOS_EU" => new(system: VSystemID.Raw.GEN, firmware: "CD_BIOS_EU"),
+					"CD_BIOS_JP" => new(system: VSystemID.Raw.GEN, firmware: "CD_BIOS_JP"),
+					"CD_BIOS_US" => new(system: VSystemID.Raw.GEN, firmware: "CD_BIOS_US"),
+					"GG_BIOS" => new(system: VSystemID.Raw.SMS, firmware: "Japan"),
+					"MS_BIOS_EU" => new(system: VSystemID.Raw.SMS, firmware: "Export"),
+					"MS_BIOS_JP" => new(system: VSystemID.Raw.SMS, firmware: "Japan"),
+					"MS_BIOS_US" => new(system: VSystemID.Raw.SMS, firmware: "Export"),
+					_ => null
+				};
				if (firmwareID != null)
				{
					// this path will be the most common PEBKAC error, so be a bit more vocal about the problem
-					srcdata = CoreComm.CoreFileProvider.GetFirmware(new(system, firmwareID), "GPGX firmwares are usually required.");
+					srcdata = CoreComm.CoreFileProvider.GetFirmware(firmwareID, "GPGX firmwares are usually required.");
					if (srcdata == null)
					{
-						Console.WriteLine("Frontend couldn't satisfy firmware request GEN:{0}", firmwareID);
+						Console.WriteLine($"Frontend couldn't satisfy firmware request {firmwareID}");
```

File: src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs [hunk 14]
```diff
@@ -245,27 +246,25 @@ private int load_archive(string filename, IntPtr buffer, int maxsize)
...
+ 					"MS_BIOS_JP" => new(system: VSystemID.Raw.SMS, firmware: "Japan"),
+ 					"MS_BIOS_US" => new(system: VSystemID.Raw.SMS, firmware: "Export"),
+ 					_ => null
+ 				};
+
 				if (firmwareID != null)
 				{
 					// this path will be the most common PEBKAC error, so be a bit more vocal about the problem
-					srcdata = CoreComm.CoreFileProvider.GetFirmware(new(system, firmwareID), "GPGX firmwares are usually required.");
+					srcdata = CoreComm.CoreFileProvider.GetFirmware((FirmwareID)firmwareID, "GPGX firmwares are usually required.");
```
Comment (YoshiRulz) [inline src/BizHawk.Emulation.Cores/Consoles/Sega/gpgx64/GPGX.cs:264]: Whoops, it's `firmwareID.Value` to unbox, sorry.

File: src/BizHawk.Client.Common/config/Config.cs [hunk 15]
```diff
@@ -36,6 +36,10 @@ public class Config
 				new[] { CoreNames.Gambatte, CoreNames.Sameboy, CoreNames.GbHawk, CoreNames.SubGbHawk }),
 			(new[] { VSystemID.Raw.GBL },
 				new[] { CoreNames.GambatteLink, CoreNames.GBHawkLink, CoreNames.GBHawkLink3x, CoreNames.GBHawkLink4x }),
+			(new[] { VSystemID.Raw.GEN },
+				new[] { CoreNames.Gpgx, CoreNames.PicoDrive }),
+			(new[] { VSystemID.Raw.GG,VSystemID.Raw.SMS, VSystemID.Raw.SG },
```
Comment (YoshiRulz) [inline src/BizHawk.Client.Common/config/Config.cs:41]: The first in the "key" array is used as the label, so SMS should probably be first.

**Review by YoshiRulz — APPROVED** [review]
Works on my machine.

It's worth noting that this also adds PicoDrive as an option for MD/Genesis, though I can't find the feature request for that.

**Comment by vadosnaprimer** [comment]
> It's worth noting that this also adds PicoDrive as an option for MD/Genesis, though I can't find the feature request for that.

The PR title needs to be edited to indicate this, otherwise it can be missed in the changelog. @SergioMartin86

**Comment by YoshiRulz** [comment]
When you squash, you're able to edit the commit description. (Also, you can edit the PR title yourself.)

**Comment by vadosnaprimer** [comment]
For some reason selecting the Nuked core results in no FM sound. I tested in `Kenseiden (J) [!].sms`. MAME option sounds correctly.

**Comment by vadosnaprimer** [comment]
Is it possible to not show all settings for all consoles?

изображение

**Comment by YoshiRulz** [comment]
We briefly discussed that on Discord, and came up with a couple of ideas that didn't seem worth it. Either hook into PropertyGrid like MAME does, or possibly have separate `ISettable<,>` implementations which are registered in the ctor rather than as an interface on the core class.

**Comment by SergioMartin86** [comment]
> > It's worth noting that this also adds PicoDrive as an option for MD/Genesis, though I can't find the feature request for that.
> 
> The PR title needs to be edited to indicate this, otherwise it can be missed in the changelog. @SergioMartin86



> > It's worth noting that this also adds PicoDrive as an option for MD/Genesis, though I can't find the feature request for that.
> 
> The PR title needs to be edited to indicate this, otherwise it can be missed in the changelog. @SergioMartin86

PicoDrive is the current fallback emulator for Genesis. I only made it an option to prioritize it in case users may want to use it instead of GPGX.

**Pull request merged**